Drivers/Src/MotorController.c

```diff
@@ -1,124 +1,217 @@
+    CPU_TS ts;
+	OS_ERR err;
+	
+	// Check to see if a mailbox is available: BLOCKING
+	OSSemPend(
+        &MotorController_ReceiveSem4,
+	    0,    
+		OS_OPT_PEND_BLOCKING,
+		&ts,
+		&err);
+	assertOSError(0, err);
+    ErrorStatus status = BSP_CAN_Read(CAN_3, &id, data);
 
-    if(length>0){
+    if(status == SUCCESS){
         message->id = id;
         //get first number (bits 0-31)
         for(int j = 0; j < MAX_CAN_LEN/2; j++){
+            firstSum <<= 8;
             firstSum += data[j];
-            if(j != MAX_CAN_LEN/2 - 1){
-                firstSum = firstSum << 8;
-            }
         }
         //get second number (bits 32-63)
         for(int k = MAX_CAN_LEN/2; k < MAX_CAN_LEN; k++){
+            secondSum <<= 8;
             secondSum += data[k];
-            if(k != MAX_CAN_LEN-1){
-                secondSum = secondSum << 8;
-            }
         }
         message->firstNum = firstSum;
         message->secondNum = secondSum;
-
-        union {
-            uint32_t n;
-            float f;
-        } convert;
         
         switch (id) {
             // If we're reading the output from the Motor Status command (0x241) then 
             // Check the status bits we care about and set flags accordingly
             case MOTOR_STATUS: {
-                if(MASK_MOTOR_TEMP_ERR & firstSum)
-                {
-                    car->MotorErrorCode.motorTempErr = ON;
+                if(MASK_MOTOR_TEMP_ERR & firstSum){
+                    Motor_FaultBitmap |= T_TEMP_ERR;
                 }
 
-                if(MASK_SS_ERR & firstSum)
-                {
-                    car->MotorErrorCode.slipSpeedErr = ON;
+                if(MASK_SS_ERR & firstSum){
+                    Motor_FaultBitmap |= T_CC_VEL_ERR;
                 }
 
-                if(MASK_CC_ERR & firstSum)
-                {
-                    car->MotorErrorCode.CCVelocityErr = ON;
+                if(MASK_CC_ERR & firstSum){
+                    Motor_FaultBitmap |= T_SLIP_SPEED_ERR;
                 }
 
-                if(MASK_OVER_SPEED_ERR & firstSum)
-                {
-                    car->MotorErrorCode.overSpeedErr = ON;
+                if(MASK_OVER_SPEED_ERR & firstSum){
+                    Motor_FaultBitmap |= T_OVER_SPEED_ERR;
                 }
+
+                assertTritiumError(Motor_FaultBitmap);
+                
                 break;
             }
-            case MOTOR_DRIVE: {
-                convert.n = secondSum;
-                car->CurrentVelocity = convert.f;
+            case MOTOR_VELOCITY: {
+                CurrentVelocity = *(float *) &secondSum;
                 break;
             }
             default: break;
         }
 
+
         return SUCCESS;
     }
     return ERROR;
 }
+
+/**
+ * @brief   Mutex protected read from Velocity parameter
+ * @return  velocity value obtained from MotorController_Read
+ */ 
+float MotorController_ReadVelocity(void){
+    return CurrentVelocity;
+}
```

Comment:
Not sure this is actually necessary, I think we'll end up doing a motor controller read every time and just using the values returned from that, but we can leave this in. Whoever writes updateVelocity may use this for convenience instead of manually unpacking the data.